### PR TITLE
Create Namespace Limiter for Pools

### PR DESCRIPTION
**Pull request by maknop**
Adding a limit for the number of total namespaces allowed to be up and running. This will only apply to certain pools. The need for this was first decided for the managed-kafka pool.

### CODE REVIEW

File: controllers/cloud.redhat.com/namespacereservation_controller.go [hunk 1]
```diff
@@ -221,6 +223,8 @@ func (r *NamespaceReservationReconciler) Reconcile(ctx context.Context, req ctrl
 func (r *NamespaceReservationReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&crd.NamespaceReservation{}).
+		Watches(&source.Kind{Type: &core.Namespace{}},
```
Comment (psav) [inline controllers/cloud.redhat.com/namespacereservation_controller.go:226]: Does the namespace reservation need to be triggered on changes to a namespace it owns? It seems like without extra logic in this controller it doesn't look like this addition performs anything new.

File: controllers/cloud.redhat.com/pool_controller.go [hunk 2]
```diff
@@ -162,6 +208,12 @@ func (r *NamespacePoolReconciler) checkReadyNamespaceQuantity(pool crd.Namespace
 	size := pool.Spec.Size
 	ready := pool.Status.Ready
 	creating := pool.Status.Creating
+	reserved := pool.Status.Reserved
+
+	if ready+creating+reserved == pool.Spec.SizeLimit && pool.Spec.SizeLimit != 0 {
+		r.Log.Info(fmt.Sprintf("Max number of namespaces already created [%d] for pool [%s]", pool.Spec.SizeLimit, pool.Name))
+		return 0
```
Comment (psav) [inline controllers/cloud.redhat.com/pool_controller.go:215]: The function is called `checkReadyNamespaceQuantity`, yet we are returning 0, in a situation where we are out of spaces to create new ones, and later on, the number of available slots, though this also doesn't seem to take into account reserved. I think this should be checked.

**Review by psav — CHANGES_REQUESTED** [review]
Still needs a little work and would be great to get some tests on this, though I'm happy if the functionality is there, to defer this to the tech debt period coming up shortly.

File: controllers/cloud.redhat.com/pool_controller.go (hunk 2 above)
Comment (maknop) [reply controllers/cloud.redhat.com/pool_controller.go:215]: I think I resolved the ask here but let me know :smile:

**Comment by maknop** [comment]
I'll add the test to the tech debt sprint :smile:

**Pull request closed without merging**